<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform the scMerge algorithm — scMerge • scMerge</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform the scMerge algorithm — scMerge"><meta property="og:description" content="Merge single-cell RNA-seq data from different batches and experiments 
leveraging (pseudo)-replicates and control genes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMerge</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.9.99</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/scMerge.html">scMerge</a>
</li>
<li>
  <a href="../articles/scMerge2.html">scMerge2</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/SydneyBioX/scMerge" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform the scMerge algorithm</h1>
    <small class="dont-index">Source: <a href="https://github.com/SydneyBioX/scMerge/blob/HEAD/R/scMerge.R" class="external-link"><code>R/scMerge.R</code></a></small>
    <div class="hidden name"><code>scMerge.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Merge single-cell RNA-seq data from different batches and experiments 
leveraging (pseudo)-replicates and control genes.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scMerge</span><span class="op">(</span></span>
<span>  <span class="va">sce_combine</span>,</span>
<span>  ctl <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exprs <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  hvg_exprs <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  batch_name <span class="op">=</span> <span class="st">"batch"</span>,</span>
<span>  marker <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  marker_list <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ruvK <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  replicate_prop <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  cell_type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell_type_match <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cell_type_inc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  BSPARAM <span class="op">=</span> <span class="fu">ExactParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  svd_k <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  dist <span class="op">=</span> <span class="st">"cor"</span>,</span>
<span>  WV <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  WV_marker <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  return_all_RUV <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_igraph <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sce_combine</dt>
<dd><p>A <code>SingleCellExperiment</code> object contains the batch-combined matrix with batch info in colData.</p></dd>


<dt>ctl</dt>
<dd><p>A character vector of negative control. It should have a non-empty intersection with the rows of sce_combine.</p></dd>


<dt>kmeansK</dt>
<dd><p>A vector indicates the kmeans's K for each batch. The length of kmeansK needs to be the same as the number of batch.</p></dd>


<dt>exprs</dt>
<dd><p>A string indicating the name of the assay requiring batch correction in sce_combine, default is logcounts.</p></dd>


<dt>hvg_exprs</dt>
<dd><p>A string indicating the assay that to be used for highly variable genes identification in sce_combine, default is counts.</p></dd>


<dt>batch_name</dt>
<dd><p>A character indicating the name of the batch column, default to "batch"</p></dd>


<dt>marker</dt>
<dd><p>An optional vector of markers, to be used in calculation of mutual nearest cluster. If no markers input, highly variable genes will be used instead.</p></dd>


<dt>marker_list</dt>
<dd><p>An optional list of markers for each batch, which will be used in calculation of mutual nearest cluster.</p></dd>


<dt>ruvK</dt>
<dd><p>An optional integer/vector indicating the number of unwanted variation factors that are removed, default is 20.</p></dd>


<dt>replicate_prop</dt>
<dd><p>A number indicating the ratio of cells that are included in pseudo-replicates, ranges from 0 to 1. Default to 1.</p></dd>


<dt>cell_type</dt>
<dd><p>An optional vector indicating the cell type information for each cell in the batch-combined matrix. If it is <code>NULL</code>, pseudo-replicate procedure will be run to identify cell type.</p></dd>


<dt>cell_type_match</dt>
<dd><p>An optional logical input for whether to find mutual nearest cluster using cell type information.</p></dd>


<dt>cell_type_inc</dt>
<dd><p>An optional vector indicating the indices of the cells that will be used to supervise the pseudo-replicate procedure.</p></dd>


<dt>BSPARAM</dt>
<dd><p>A <code>BiocSingularParam</code> class object from the <code>BiocSingular</code> package is used. Default is ExactParam().</p></dd>


<dt>svd_k</dt>
<dd><p>If BSPARAM is set to <code>RandomParam</code> or <code>IrlbaParam</code> class from <code>BiocSingular</code> package, then 
<code>svd_k</code> will be used to used to reduce the computational cost of singular value decomposition. Default to 50.</p></dd>


<dt>dist</dt>
<dd><p>The distance metrics that are used in the calculation of the mutual nearest cluster, default is Pearson correlation.</p></dd>


<dt>WV</dt>
<dd><p>A optional vector indicating the wanted variation factor other than cell type info, such as cell stages.</p></dd>


<dt>WV_marker</dt>
<dd><p>An optional vector indicating the markers of the wanted variation.</p></dd>


<dt>BPPARAM</dt>
<dd><p>A <code>BiocParallelParam</code> class object from the <code>BiocParallel</code> package is used. Default is SerialParam().</p></dd>


<dt>return_all_RUV</dt>
<dd><p>If <code>FALSE</code>, then only returns a <code>SingleCellExperiment</code> object with original data and one normalised matrix.
Otherwise, the <code>SingleCellExperiment</code> object will contain the original data and one normalised matrix for <code>each</code> ruvK value. In this latter case, assay_name must have the same length as ruvK.</p></dd>


<dt>assay_name</dt>
<dd><p>The assay name(s) for the adjusted expression matrix(matrices). If <code>return_all_RUV = TRUE</code> assay_name must have the same length as ruvK.</p></dd>


<dt>plot_igraph</dt>
<dd><p>If <code>TRUE</code>, then during the un/semi-supervised scMerge, igraph plot will be displayed</p></dd>


<dt>verbose</dt>
<dd><p>If <code>TRUE</code>, then all intermediate steps will be shown. Default to <code>FALSE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Returns a <code>SingleCellExperiment</code> object with following components:</p><ul><li><p>assays: the original assays and also the normalised matrix</p></li>
<li><p>metadata: containing the ruvK vector, ruvK_optimal based on F-score, and the replicate matrix</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Yingxin Lin, Kevin Wang</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Loading example data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'example_sce'</span>, package <span class="op">=</span> <span class="st">'scMerge'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## Previously computed stably expressed genes</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'segList_ensemblGeneID'</span>, package <span class="op">=</span> <span class="st">'scMerge'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## Running an example data with minimal inputs</span></span></span>
<span class="r-in"><span><span class="va">sce_mESC</span> <span class="op">&lt;-</span> <span class="fu">scMerge</span><span class="op">(</span>sce_combine <span class="op">=</span> <span class="va">example_sce</span>,</span></span>
<span class="r-in"><span>ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span></span>
<span class="r-in"><span>kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>assay_name <span class="op">=</span> <span class="st">'scMerge'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dimension of the replicates mapping matrix: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 200   3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Step 2: Performing RUV normalisation. This will take minutes to hours. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> scMerge complete!</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sce_mESC</span> <span class="op">=</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce_mESC</span>, exprs_values <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span>                      </span></span>
<span class="r-in"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce_mESC</span>, colour_by <span class="op">=</span> <span class="st">'cellTypes'</span>, shape <span class="op">=</span> <span class="st">'batch'</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="scMerge-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sce_mESC</span> <span class="op">=</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce_mESC</span>, exprs_values <span class="op">=</span> <span class="st">'scMerge'</span><span class="op">)</span>                                       </span></span>
<span class="r-in"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sce_mESC</span>, colour_by <span class="op">=</span> <span class="st">'cellTypes'</span>, shape <span class="op">=</span> <span class="st">'batch'</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="scMerge-2.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yingxin Lin, Kevin Wang, Sydney Bioinformatics and Biometrics Group.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

