<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create replicate matrix for scMerge algorithm — scReplicate • scMerge</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Create replicate matrix for scMerge algorithm — scReplicate" />
<meta property="og:description" content="Create replicate matrix for scMerge algorithm using un-/semi-/supervised approaches." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMerge</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/scMerge.html">Vignette</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Case Study
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://sydneybiox.github.io/scMerge/articles/case_study/Breast_Data.html">Breast cancer data</a>
    </li>
    <li>
      <a href="https://sydneybiox.github.io/scMerge/articles/case_study/CellBench_Data.html">CellBench data</a>
    </li>
    <li>
      <a href="https://sydneybiox.github.io/scMerge/articles/case_study/ESC_Data.html">Embyronic stem cell data</a>
    </li>
    <li>
      <a href="https://sydneybiox.github.io/scMerge/articles/case_study/Mouse_Liver_Data.html">Mouse liver data</a>
    </li>
    <li>
      <a href="https://sydneybiox.github.io/scMerge/articles/case_study/Neuronal_Data.html">Neuronal single cell data</a>
    </li>
    <li>
      <a href="https://sydneybiox.github.io/scMerge/articles/case_study/Pancreas4_Data.html">Four pancreas data</a>
    </li>
    <li>
      <a href="https://sydneybiox.github.io/scMerge/articles/case_study/mESC_Data.html">Mouse Embryonic Stem Cell Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/SydneyBioX/scMerge">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create replicate matrix for scMerge algorithm</h1>
    <small class="dont-index">Source: <a href='https://github.com/SydneyBioX/scMerge/blob/master/R/scReplicate.R'><code>R/scReplicate.R</code></a></small>
    <div class="hidden name"><code>scReplicate.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Create replicate matrix for scMerge algorithm using un-/semi-/supervised approaches.</p>
    </div>

    <pre class="usage"><span class='fu'>scReplicate</span>(
  <span class='kw'>sce_combine</span>,
  batch = <span class='kw'>NULL</span>,
  kmeansK = <span class='kw'>NULL</span>,
  exprs = <span class='st'>"logcounts"</span>,
  hvg_exprs = <span class='st'>"counts"</span>,
  marker = <span class='kw'>NULL</span>,
  marker_list = <span class='kw'>NULL</span>,
  replicate_prop = <span class='fl'>1</span>,
  cell_type = <span class='kw'>NULL</span>,
  cell_type_match = <span class='fl'>FALSE</span>,
  cell_type_inc = <span class='kw'>NULL</span>,
  dist = <span class='st'>"cor"</span>,
  WV = <span class='kw'>NULL</span>,
  WV_marker = <span class='kw'>NULL</span>,
  BPPARAM = <span class='fu'>SerialParam</span>(),
  return_all = <span class='fl'>FALSE</span>,
  BSPARAM = <span class='fu'>ExactParam</span>(),
  plot_igraph = <span class='fl'>TRUE</span>,
  verbose = <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sce_combine</th>
      <td><p>A <code>SingleCellExperiment</code> object contains the batch-combined matrix with batch info in colData</p></td>
    </tr>
    <tr>
      <th>batch</th>
      <td><p>A vector indicates the batch information for each cell in the batch-combined matrix.</p></td>
    </tr>
    <tr>
      <th>kmeansK</th>
      <td><p>A vector indicates the kmeans's K for each batch, length of kmeansK needs to be the same as the number of batch.</p></td>
    </tr>
    <tr>
      <th>exprs</th>
      <td><p>A string indicates the assay that are used for batch correction, default is logcounts</p></td>
    </tr>
    <tr>
      <th>hvg_exprs</th>
      <td><p>A string indicates the assay that are used for highly variable genes identification, default is counts</p></td>
    </tr>
    <tr>
      <th>marker</th>
      <td><p>A vector of markers, which will be used in calculation of mutual nearest cluster. If no markers input, highly variable genes will be used instead</p></td>
    </tr>
    <tr>
      <th>marker_list</th>
      <td><p>A list of markers for each batch, which will be used in calculation of mutual nearest cluster.</p></td>
    </tr>
    <tr>
      <th>replicate_prop</th>
      <td><p>A number indicating the ratio of cells that are included in pseudo-replicates, ranges from 0 to 1. Default to 1.</p></td>
    </tr>
    <tr>
      <th>cell_type</th>
      <td><p>A vector indicates the cell type information for each cell in the batch-combined matrix. If it is <code>NULL</code>, pseudo-replicate procedure will be run to identify cell type.</p></td>
    </tr>
    <tr>
      <th>cell_type_match</th>
      <td><p>Whether find mutual nearest cluster using cell type information</p></td>
    </tr>
    <tr>
      <th>cell_type_inc</th>
      <td><p>A vector indicates the indices of the cells that will be used to supervise the pseudo-replicate procedure</p></td>
    </tr>
    <tr>
      <th>dist</th>
      <td><p>The distance metrics that are used in the calculation of the mutual nearest cluster, default is Pearson correlation.</p></td>
    </tr>
    <tr>
      <th>WV</th>
      <td><p>A vector indicates the wanted variation factor other than cell type info, such as cell stages.</p></td>
    </tr>
    <tr>
      <th>WV_marker</th>
      <td><p>A vector indicates the markers of the wanted variation.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>A <code>BiocParallelParam</code> class object from the <code>BiocParallel</code> package is used. Default is SerialParam().</p></td>
    </tr>
    <tr>
      <th>return_all</th>
      <td><p>If <code>FALSE</code>, only return the replicate matrix.</p></td>
    </tr>
    <tr>
      <th>BSPARAM</th>
      <td><p>A <code>BiocSingularParam</code> class object from the <code>BiocSingular</code> package is used. Default is ExactParam().</p></td>
    </tr>
    <tr>
      <th>plot_igraph</th>
      <td><p>If <code>TRUE</code>, then during the un/semi-supervised scMErge, igraph plot will be displayed</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>If <code>TRUE</code>, then all intermediate steps will be shown. Default to <code>FALSE</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>If <code>return_all</code> is <code>FALSE</code>, return a replicate matrix.
If <code>return_sce</code> is <code>TRUE</code>, return the followings</p>
<dt>repMat </dt><dd><p>replicate matrix</p></dd>
<dt>mnc </dt><dd><p>mutual nearest cluster</p></dd>
<dt>replicate vector </dt><dd><p>replicate vector</p></dd>
<dt>HVG </dt><dd><p>highly variable genes used in scReplicate</p></dd>

A cell-replicates mapping matrix.
Each row correspond to a cell from the input expression matrix, and each column correspond to a cell-cluster/cell-type.
An element of the mapping matrix is 1 if the scReplicate algorithm determines that this cell
should belong to that cell cluster and 0 otherwise.

    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Yingxin Lin, Kevin Wang</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Loading example data</span>
<span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span>(<span class='fl'>1</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>'example_sce'</span>, package = <span class='st'>'scMerge'</span>)
<span class='kw'>scRep_result</span> <span class='op'>=</span> <span class='fu'>scReplicate</span>(
  sce_combine = <span class='kw'>example_sce</span>,
  batch = <span class='kw'>example_sce</span><span class='op'>$</span><span class='kw'>batch</span>,
  kmeansK = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>3</span>,<span class='fl'>3</span>))</div><div class='img'><img src='scReplicate-1.png' alt='' width='700' height='433' /></div><div class='input'></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Yingxin Lin, Kevin Wang, Sydney Bioinformatics and Biometrics Group.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


