<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scMerge: Merging multiple batches of scRNA-seq data â€¢ scMerge</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="scMerge: Merging multiple batches of scRNA-seq data">
<meta property="og:description" content="Like all gene expression data, single-cell RNA-seq (scRNA-Seq) data suffers from 
    batch effects and other unwanted variations that makes accurate biological interpretations difficult. 
    The scMerge method leverages stably expressed genes (SEGs) and (pseudo-) replicates to 
    identify different sources of variations so they can be removed from the raw scRNA-Seq data. 
    This package contains all the necessary functions in the 
    scMerge pipeline, including the identification of SEGs, replication-identification methods, and 
    removal of unwanted variations in scRNA-Seq data. ">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scMerge</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/scMerge.html">Vignette</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Case Study
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Breast_Data/Breast_Data.html">Breast Cancer Data</a>
    </li>
    <li>
      <a href="articles/CellBench_Data/CellBench_Data.html">CellBench Data</a>
    </li>
    <li>
      <a href="articles/ESC_Data/ESC_Data.html">Embyronic Stem Cell Data</a>
    </li>
    <li>
      <a href="articles/Mouse_Liver_Data/Mouse_Liver_Data.html">Mouse Liver Data</a>
    </li>
    <li>
      <a href="articles/Neuronal_Data/Neuronal_Data.html">Neuronal Single Cell Data</a>
    </li>
    <li>
      <a href="articles/Pancreas4_Data/Pancreas4_Data.html">Pancreas 4 Data</a>
    </li>
    <li>
      <a href="articles/mESC_Data/mESC_Data.html">Mouse Embryonic Stem Cell Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SydneyBioX/scMerge">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="scmerge" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#scmerge" class="anchor"></a>scMerge <img src="../../../../../Dropbox%20(Sydney%20Uni)/Single%20Cell%20Reserach/SCProj-Normalization/codes/scMerge/man/figures/logo.png" align="right" width="200">
</h1></div>


<p><code>scMerge</code> is a R package for merging and normalising single-cell RNA-Seq datasets.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>The installation process could take up to 5 minutes, depending if you have some of the packages pre-installed.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co">## Some CRAN packages required by scMerge</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"BiocManager"</span>, <span class="st">"cluster"</span>, <span class="st">"distr"</span>, <span class="st">"doSNOW"</span>, </a>
<a class="sourceLine" id="cb1-3" title="3">                   <span class="st">"foreach"</span>, <span class="st">"igraph"</span>, <span class="st">"irlba"</span>, <span class="st">"iterators"</span>, </a>
<a class="sourceLine" id="cb1-4" title="4">                   <span class="st">"matrixStats"</span>, <span class="st">"pdist"</span>, <span class="st">"proxy"</span>,  <span class="st">"Rcpp"</span>, </a>
<a class="sourceLine" id="cb1-5" title="5">                   <span class="st">"RcppEigen"</span>, <span class="st">"rsvd"</span>, <span class="st">"ruv"</span>))</a>
<a class="sourceLine" id="cb1-6" title="6"></a>
<a class="sourceLine" id="cb1-7" title="7"><span class="co">## Some BioConductor packages required by scMerge</span></a>
<a class="sourceLine" id="cb1-8" title="8">BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"BiocParallel"</span>, <span class="st">"M3Drop"</span>, <span class="st">"SingleCellExperiment"</span>))</a>
<a class="sourceLine" id="cb1-9" title="9"></a>
<a class="sourceLine" id="cb1-10" title="10"><span class="co">## Install scMerge from BioConductor, requires R 3.6.0 or above</span></a>
<a class="sourceLine" id="cb1-11" title="11">BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"scMerge"</span>, <span class="dt">version =</span> <span class="st">"3.9"</span>)</a>
<a class="sourceLine" id="cb1-12" title="12"></a>
<a class="sourceLine" id="cb1-13" title="13"></a>
<a class="sourceLine" id="cb1-14" title="14"><span class="co">## Install the development version of scMerge</span></a>
<a class="sourceLine" id="cb1-15" title="15">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"SydneyBioX/scMerge"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="vignette" class="section level2">
<h2 class="hasAnchor">
<a href="#vignette" class="anchor"></a>Vignette</h2>
<p>You can find the vignette at our website: <a href="https://sydneybiox.github.io/scMerge/index.html" class="uri">https://sydneybiox.github.io/scMerge/index.html</a>.</p>
</div>
<div id="case-studies" class="section level2">
<h2 class="hasAnchor">
<a href="#case-studies" class="anchor"></a>Case studies</h2>
<p>You can find a list of case studies here: <a href="https://sydneybiox.github.io/scMerge/articles/" class="uri">https://sydneybiox.github.io/scMerge/articles/</a>.</p>
</div>
<div id="contact-us" class="section level2">
<h2 class="hasAnchor">
<a href="#contact-us" class="anchor"></a>Contact us</h2>
<p>If you have any enquires, especially about performing <code>scMerge</code> integration on your data, then please contact <a href="mailto:bioinformatics@maths.usyd.edu.au" class="email">bioinformatics@maths.usyd.edu.au</a>.</p>
</div>
<div id="reference" class="section level2">
<h2 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h2>
<p><strong>scMerge: Integration of multiple single-cell transcriptomics datasets leveraging stable expression and pseudo-replication</strong></p>
<p>Yingxin Lin, Shila Ghazanfar, Kevin Y.X. Wang, Johann A. Gagnon-Bartsch, Kitty K. Lo, Xianbin Su, Ze-Guang Han, John T. Ormerod, Terence P. Speed, Pengyi Yang, Jean Y. H. Yang</p>
<p>doi: <a href="https://doi.org/10.1101/393280" class="uri">https://doi.org/10.1101/393280</a></p>
<p>BioRxiv preprint: <a href="https://www.biorxiv.org/content/early/2018/08/16/393280" class="uri">https://www.biorxiv.org/content/early/2018/08/16/393280</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/SydneyBioX/scMerge">https://â€‹github.com/â€‹SydneyBioX/â€‹scMerge</a>
</li>
<li>Report a bug at <br><a href="https://github.com/SydneyBioX/scMerge/issues">https://â€‹github.com/â€‹SydneyBioX/â€‹scMerge/â€‹issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing scMerge</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Kevin Wang <br><small class="roles"> Author, maintainer </small>  </li>
<li>Yingxin Lin <br><small class="roles"> Author </small>  </li>
<li>Sydney Bioinformatics and Biometrics Group <br><small class="roles"> Funder </small>  </li>
</ul>
</div>

      <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/SydneyBioX/scMerge"><img src="https://travis-ci.org/SydneyBioX/scMerge.svg?branch=master" alt="Travis build status"></a></li>
<li><a href="https://codecov.io/gh/SydneyBioX/scMerge?branch=master"><img src="https://codecov.io/gh/SydneyBioX/scMerge/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by Kevin Wang, Yingxin Lin, Sydney Bioinformatics and Biometrics Group.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
