<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to the scMerge package • scMerge</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to the scMerge package">
<meta property="og:description" content="scMerge">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMerge</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.19.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/scMerge.html">scMerge</a>
</li>
<li>
  <a href="../articles/scMerge2.html">scMerge2</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SydneyBioX/scMerge" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to the scMerge package</h1>
                        <h4 data-toc-skip class="author">Yingxin
Lin</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, The University of Sydney,
Australia<br><h4 data-toc-skip class="author">Kevin
Y.X. Wang</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, The University of Sydney,
Australia<br><small class="dont-index">Source: <a href="https://github.com/SydneyBioX/scMerge/blob/HEAD/vignettes/scMerge.Rmd" class="external-link"><code>vignettes/scMerge.Rmd</code></a></small>
      <div class="hidden name"><code>scMerge.Rmd</code></div>

    </address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The scMerge algorithm allows batch effect removal and normalisation
for single cell RNA-Seq data. It comprises of three key components
including:</p>
<ol style="list-style-type: decimal">
<li>The identification of stably expressed genes (SEGs) as “negative
controls” for estimating unwanted factors;</li>
<li>The construction of pseudo-replicates to estimate the effects of
unwanted factors; and</li>
<li>The adjustment of the datasets with unwanted variation using a
fastRUVIII model.</li>
</ol>
<p>The purpose of this vignette is to illustrate some uses of
<code>scMerge</code> and explain its key components.</p>
</div>
<div class="section level2">
<h2 id="loading-packages-and-data">Loading Packages and Data<a class="anchor" aria-label="anchor" href="#loading-packages-and-data"></a>
</h2>
<p>We will load the <code>scMerge</code> package. We designed our
package to be consistent with the popular BioConductor’s single cell
analysis framework, namely the <code>SingleCellExperiment</code> and
<code>scater</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SydneyBioX/scMerge" class="external-link">scMerge</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>We provided an illustrative mouse embryonic stem cell (mESC) data in
our package, as well as a set of pre-computed stably expressed gene
(SEG) list to be used as negative control genes.</p>
<p>The full curated, unnormalised mESC data can be found <a href="http://www.maths.usyd.edu.au/u/yingxinl/wwwnb/scMergeData/sce_mESC.rda" class="external-link">here</a>.
The <code>scMerge</code> package comes with a sub-sampled, two-batches
version of this data (named “batch2” and “batch3” to be consistent with
the full data) .</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subsetted mouse ESC data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_sce"</span>, package <span class="op">=</span> <span class="st">"scMerge"</span><span class="op">)</span></span></code></pre></div>
<p>In this mESC data, we pooled data from 2 different batches from three
different cell types. Using a PCA plot, we can see that despite strong
separation of cell types, there is also a strong separation due to batch
effects. This information is stored in the <code>colData</code> of
<code>example_sce</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">example_sce</span>, exprs_values <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span></span>
<span>  <span class="va">example_sce</span>, </span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellTypes"</span>, </span>
<span>  shape_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scMerge_files/figure-html/checking%20raw%20data-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="illustrating-pseudo-replicates-constructions">Illustrating pseudo-replicates constructions<a class="anchor" aria-label="anchor" href="#illustrating-pseudo-replicates-constructions"></a>
</h2>
<p>The first major component of <code>scMerge</code> is to obtain
negative controls for our normalisation. In this vignette, we will be
using a set of pre-computed SEGs from a single cell mouse data made
available through the <code>segList_ensemblGeneID</code> data in our
package. For more information about the selection of negative controls
and SEGs, please see Section <a href="#selectnc">select SEGs</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## single-cell stably expressed gene list</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"segList_ensemblGeneID"</span>, package <span class="op">=</span> <span class="st">"scMerge"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ENSMUSG00000058835" "ENSMUSG00000026842" "ENSMUSG00000027671"</span></span>
<span><span class="co">#&gt; [4] "ENSMUSG00000020152" "ENSMUSG00000054693" "ENSMUSG00000049470"</span></span></code></pre></div>
<p>The second major component of <code>scMerge</code> is to compute
pseudo-replicates for cells so we can perform normalisation. We offer
three major ways of computing this pseudo-replicate information:</p>
<ol style="list-style-type: decimal">
<li>Unsupervised clustering, using k-means clustering;</li>
<li>Supervised clustering, using known cell type information; and</li>
<li>Semi-supervised clustering, using partially known cell type
information.</li>
</ol>
</div>
<div class="section level2">
<h2 id="unsupervised-scmerge">Unsupervised <code>scMerge</code><a class="anchor" aria-label="anchor" href="#unsupervised-scmerge"></a>
</h2>
<p>In unsupervised <code>scMerge</code>, we will perform a k-means
clustering to obtain pseudo-replicates. This requires the users to
supply a <code>kmeansK</code> vector with each element indicating number
of clusters in each of the batches. For example, we know “batch2” and
“batch3” both contain three cell types. Hence,
<code>kmeansK = c(3, 3)</code> in this case.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_unsupervised</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">example_sce</span>, </span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_unsupervised"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Step 2: Performing RUV normalisation. This will take minutes to hours.</span></span>
<span><span class="co">#&gt; scMerge complete!</span></span></code></pre></div>
<p>We now colour construct the PCA plot again on our normalised data. We
can observe a much better separation by cell type and less separation by
batches.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_unsupervised</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">scMerge_unsupervised</span>, exprs_values <span class="op">=</span> <span class="st">"scMerge_unsupervised"</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span></span>
<span>  <span class="va">scMerge_unsupervised</span>, </span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellTypes"</span>, </span>
<span>  shape_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scMerge_files/figure-html/unsupervised_default_plotting-1.png" width="700"></p>
<!-- ##Selecting 80% of cells -->
<!-- ```{r results='hide',fig.show='hide'} -->
<!-- system.time(sce_mESC <- scMerge(sce_mESC,  -->
<!--                     ctl = segList_ensemblGeneID$mouse$mouse_scSEG, -->
<!--                     kmeansK = c(1,3,3,1,1), -->
<!--                     assay_name = "scMerge_unsupervised_80", -->
<!--                     replicate_prop = 0.8)) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- sce_mESC <- runPCA(sce_mESC, exprs_values = "scMerge_unsupervised_80") -->
<!-- scater::plotPCA(sce_mESC, colour_by="cellTypes",shape_by="batch") -->
<!-- ``` -->
</div>
<div class="section level2">
<h2 id="selecting-all-cells">Selecting all cells<a class="anchor" aria-label="anchor" href="#selecting-all-cells"></a>
</h2>
<p>By default, <code>scMerge</code> only uses 50% of the cells to
perform kmeans clustering. While this is sufficient to perform a
satisfactory normalisation in most cases, users can control if they wish
all cells be used in the kmeans clustering.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_unsupervised_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">example_sce</span>, </span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_unsupervised_all"</span>,</span>
<span>  replicate_prop <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Step 2: Performing RUV normalisation. This will take minutes to hours.</span></span>
<span><span class="co">#&gt; scMerge complete!</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_unsupervised_all</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">scMerge_unsupervised_all</span>,</span>
<span>                                  exprs_values <span class="op">=</span> <span class="st">"scMerge_unsupervised_all"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span></span>
<span>  <span class="va">scMerge_unsupervised_all</span>, </span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellTypes"</span>, </span>
<span>  shape_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scMerge_files/figure-html/unsupervised_prop1_plotting-1.png" width="700"></p>
<!-- ##Selecting 20% of cells -->
<!-- ```{r results='hide',fig.show='hide'} -->
<!-- system.time(sce_mESC <- scMerge(sce_mESC,  -->
<!--                     ctl = segList_ensemblGeneID$mouse$mouse_scSEG, -->
<!--                     kmeansK = c(1,3,3,1,1), -->
<!--                     assay_name = "scMerge_unsupervised_50", -->
<!--                     replicate_prop = 0.2)) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- sce_mESC <- runPCA(sce_mESC, exprs_values = "scMerge_unsupervised_50") -->
<!-- scater::plotPCA(sce_mESC, colour_by="cellTypes",shape_by="batch") -->
<!-- ``` -->
</div>
<div class="section level2">
<h2 id="supervised-scmerge">Supervised <code>scMerge</code><a class="anchor" aria-label="anchor" href="#supervised-scmerge"></a>
</h2>
<p>If <strong>all</strong> cell type information is available to the
user, then it is possible to use this information to create
pseudo-replicates. This can be done through the <code>cell_type</code>
argument in the <code>scMerge</code> function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_supervised</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_supervised"</span>,</span>
<span>  cell_type <span class="op">=</span> <span class="va">example_sce</span><span class="op">$</span><span class="va">cellTypes</span><span class="op">)</span></span>
<span><span class="co">#&gt; Step 2: Performing RUV normalisation. This will take minutes to hours.</span></span>
<span><span class="co">#&gt; scMerge complete!</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_supervised</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">scMerge_supervised</span>,</span>
<span>                            exprs_values <span class="op">=</span> <span class="st">"scMerge_supervised"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span></span>
<span>  <span class="va">scMerge_supervised</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellTypes"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scMerge_files/figure-html/supervised_plotting-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="semi-supervised-scmerge-i">Semi-supervised scMerge I<a class="anchor" aria-label="anchor" href="#semi-supervised-scmerge-i"></a>
</h2>
<p>If the user is only able to access <strong>partial</strong> cell type
information, then it is still possible to use this information to create
pseudo-replicates. This can be done through the <code>cell_type</code>
and <code>cell_type_inc</code> arguments in the <code>scMerge</code>
function. <code>cell_type_inc</code> should contain a vector of indices
indicating which elements in the <code>cell_type</code> vector should be
used to perform semi-supervised scMerge.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_semisupervised1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_semisupervised1"</span>,</span>
<span>  cell_type <span class="op">=</span> <span class="va">example_sce</span><span class="op">$</span><span class="va">cellTypes</span>,</span>
<span>  cell_type_inc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">$</span><span class="va">cellTypes</span> <span class="op">==</span> <span class="st">"2i"</span><span class="op">)</span>, </span>
<span>  cell_type_match <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Step 2: Performing RUV normalisation. This will take minutes to hours.</span></span>
<span><span class="co">#&gt; scMerge complete!</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_semisupervised1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">scMerge_semisupervised1</span>,</span>
<span>                                 exprs_values <span class="op">=</span> <span class="st">"scMerge_semisupervised1"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span></span>
<span>  <span class="va">scMerge_semisupervised1</span>, </span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellTypes"</span>, </span>
<span>  shape_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scMerge_files/figure-html/semi_supervised1_plotting-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="semi-supervised-scmerge-ii">Semi-supervised scMerge II<a class="anchor" aria-label="anchor" href="#semi-supervised-scmerge-ii"></a>
</h2>
<!-- Perform scMerge using known cell type information to identify mutual nearest cluster -->
<p>There is alternative semi-supervised method to create
pseudo-replicates for <code>scMerge</code>. This uses known cell type
information to identify mutual nearest clusters and it is achieved via
the <code>cell_type</code> and <code>cell_type_match = TRUE</code>
options in the <code>scMerge</code> function.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_semisupervised2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_semisupervised2"</span>,</span>
<span>  cell_type <span class="op">=</span> <span class="va">example_sce</span><span class="op">$</span><span class="va">cellTypes</span>,</span>
<span>  cell_type_inc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell_type_match <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Step 2: Performing RUV normalisation. This will take minutes to hours.</span></span>
<span><span class="co">#&gt; scMerge complete!</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMerge_semisupervised2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">scMerge_semisupervised2</span>,</span>
<span>                                 exprs_values <span class="op">=</span> <span class="st">"scMerge_semisupervised2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span></span>
<span>  <span class="va">scMerge_semisupervised2</span>, </span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellTypes"</span>, </span>
<span>  shape_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scMerge_files/figure-html/semi_supervised2_plotting-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="selectnc">Selecting negative controls<a class="anchor" aria-label="anchor" href="#selectnc"></a>
</h2>
<p>In simple terms, a negative control is a gene that has expression
values relatively constant across these datasets. The concept of using
these negative control genes for normalisation was most widely used in
the RUV method family (e.g. <a href="https://academic.oup.com/biostatistics/article/13/3/539/248166" class="external-link">Gagnon-Bartsch
&amp; Speed (2012)</a> and <a href="https://www.nature.com/articles/nbt.2931" class="external-link">Risso et.
al. (2014)</a>) and there exist multiple methods to find these negative
controls. In our paper, we recommened the SEGs as negative controls for
scRNA-Seq data and SEGs can be found using either a data-adaptive
computational method or external knowledge.</p>
<ul>
<li>Computation method: We provide the function <code>scSEGIndex</code>
to calculate the SEG from a data matrix. The output of this function is
a <code>data.frame</code> with a SEG index calculated for each gene. See
<a href="https://www.biorxiv.org/content/10.1101/229815v2" class="external-link">Lin et.
al. (2018)</a> for more details.</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exprs_mat</span> <span class="op">=</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">example_sce</span>, <span class="st">'counts'</span><span class="op">)</span></span>
<span><span class="va">result</span> <span class="op">=</span> <span class="fu"><a href="../reference/scSEGIndex.html">scSEGIndex</a></span><span class="op">(</span>exprs_mat <span class="op">=</span> <span class="va">exprs_mat</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>External knowledge: We have applied the SEG computational
methodology on multiple human and mouse scRNA-Seq data and made these
available as data objects in our package. The end-users can simply use
these pre-computed results. There are also additional negative controls
from bulk microarray and bulkd RNA-Seq data.</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## SEG list in ensemblGene ID</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"segList_ensemblGeneID"</span>, package <span class="op">=</span> <span class="st">"scMerge"</span><span class="op">)</span> </span>
<span><span class="co">## SEG list in official gene symbols</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"segList"</span>, package <span class="op">=</span> <span class="st">"scMerge"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## SEG list for human scRNA-Seq data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">segList</span><span class="op">$</span><span class="va">human</span><span class="op">$</span><span class="va">human_scSEG</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "AAR2"  "AATF"  "ABCF3" "ABHD2" "ABT1"  "ACAP2"</span></span>
<span></span>
<span><span class="co">## SEG list for human bulk microarray data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">segList</span><span class="op">$</span><span class="va">human</span><span class="op">$</span><span class="va">bulkMicroarrayHK</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "AATF"  "ABL1"  "ACAT2" "ACTB"  "ACTG1" "ACTN4"</span></span>
<span></span>
<span><span class="co">## SEG list for human bulk RNASeq data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">segList</span><span class="op">$</span><span class="va">human</span><span class="op">$</span><span class="va">bulkRNAseqHK</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "AAGAB"  "AAMP"   "AAR2"   "AARS"   "AARS2"  "AARSD1"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="achieving-fast-and-memory-efficient-computation">Achieving fast and memory-efficient computation<a class="anchor" aria-label="anchor" href="#achieving-fast-and-memory-efficient-computation"></a>
</h2>
<div class="section level3">
<h3 id="using-approximated-svd">Using approximated SVD<a class="anchor" aria-label="anchor" href="#using-approximated-svd"></a>
</h3>
<p>Under most circumstances, <code>scMerge</code> is fast enough to be
used on a personal laptop for a moderately large data. However, we do
recognise the difficulties associated with computation when dealing with
larger data. To this end, we devised a fast version of
<code>scMerge</code>. The major difference between the two versions lies
on the noise estimation component, which utilised singular value
decomposition (SVD). In order to speed up <code>scMerge</code>, we used
<code>BiocSingular</code> package that offers several SVD speed
improvements. This computational method is able to speed up
<code>scMerge</code> by obtain a very accurate approximation of the
noise structure in the data. This option is achieved via the option
<code>BSPARAM = IrlbaParam()</code> or
<code>BSPARAM = RandomParam()</code>. Additionally, <code>svd_k</code>
is a parameter that controlling the degree of approximations.</p>
<p>We recommend using this option in the case where the number of cells
is large in your single cell data. The speed advantage we obtain for
large single cell data is much more dramatic than on a smaller dataset
like the example mESC data. For example, a single run of normal
<code>scMerge</code> on a human <a href="https://sydneybiox.github.io/scMerge/articles/Pancreas4_Data/Pancreas4_Data.html" class="external-link">pancreas
data</a> (23699 features and 4566 cells) takes about 10 minutes whereas
the speed up version takes just under 4 minutes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/BiocSingular" class="external-link">BiocSingular</a></span><span class="op">)</span></span>
<span><span class="va">scMerge_fast</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">example_sce</span>, </span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_fast"</span>, </span>
<span>  BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>  svd_k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co">#&gt; Step 2: Performing RUV normalisation. This will take minutes to hours.</span></span>
<span><span class="co">#&gt; Warning in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth =</span></span>
<span><span class="co">#&gt; TRUE, : You're computing too large a percentage of total singular values, use a</span></span>
<span><span class="co">#&gt; standard svd instead.</span></span>
<span><span class="co">#&gt; scMerge complete!</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Normally, scMerge takes "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">t2</span> <span class="op">-</span> <span class="va">t1</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">" seconds"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Normally, scMerge takes  2.03  seconds"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Fast version of scMerge takes "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">t4</span> <span class="op">-</span> <span class="va">t3</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">" seconds"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Fast version of scMerge takes  25.75  seconds"</span></span>
<span></span>
<span><span class="va">scMerge_fast</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">scMerge_fast</span>, exprs_values <span class="op">=</span> <span class="st">"scMerge_fast"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span></span>
<span>  <span class="va">scMerge_fast</span>, </span>
<span>  colour_by <span class="op">=</span> <span class="st">"cellTypes"</span>, </span>
<span>  shape_by <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"fast scMerge yields similar results to the default version"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scMerge_files/figure-html/computation_svd_plotting-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="parallelised-computing">Parallelised computing<a class="anchor" aria-label="anchor" href="#parallelised-computing"></a>
</h3>
<p><code>scMerge</code> is implemented with a parallelised computational
option via the <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link">BiocParallel</a>
package. You can enable this option using the <code>BPPARAM</code>
argument with various <code>BiocParallelParam</code> objects that is
suitable for your operating system.</p>
<p>Please note that any parallelisation would incur a small overhead.
Hence we recommend you do not use parallelisation for small data.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="va">scMerge_parallel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">example_sce</span>, </span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_parallel"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sparse-array">Sparse array<a class="anchor" aria-label="anchor" href="#sparse-array"></a>
</h3>
<p><code>scMerge</code> also supports sparse array input, which could be
very helpful in speeding up computations and saving RAM.
<code>scMerge</code> does not perform internal matrix conversion, so you
may use the following codes as an example of converting typical
<code>matrix</code> class to sparse matrices before running
<code>scMerge</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/DelayedArray" class="external-link">DelayedArray</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">sparse_input</span> <span class="op">=</span> <span class="va">example_sce</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sparse_input</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sparse_input</span><span class="op">)</span>, <span class="st">"dgeMatrix"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sparse_input</span>, <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sparse_input</span><span class="op">)</span>, <span class="st">"dgeMatrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scMerge_sparse</span> <span class="op">=</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">sparse_input</span>, </span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_sparse"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="out-of-memory-computations-through-hdf5array">Out-of-memory computations (through <code>HDF5Array</code>)<a class="anchor" aria-label="anchor" href="#out-of-memory-computations-through-hdf5array"></a>
</h3>
<p>Bioconductor provides an infrastructure for out-of-memory computation
through <code>HDF5Array</code>. In simple terms, we can load an on-disk
data into RAM, make computations and write to hard disk. This is
particularly helpful when the data is too large for in-RAM computations.
You may use the following codes as an example of converting typical
<code>matrix</code> class to <code>HDF5Array</code> matrices before
running <code>scMerge</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/HDF5Array" class="external-link">HDF5Array</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/DelayedArray" class="external-link">DelayedArray</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DelayedArray</span><span class="fu">:::</span><span class="fu">set_verbose_block_processing</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co">## To monitor block processing </span></span>
<span></span>
<span><span class="va">hdf5_input</span> <span class="op">=</span> <span class="va">example_sce</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">hdf5_input</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">hdf5_input</span><span class="op">)</span>, <span class="st">"HDF5Array"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">hdf5_input</span>, <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">hdf5_input</span><span class="op">)</span>, <span class="st">"HDF5Array"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scMerge_hdf5</span> <span class="op">=</span> <span class="fu"><a href="../reference/scMerge.html">scMerge</a></span><span class="op">(</span></span>
<span>  sce_combine <span class="op">=</span> <span class="va">sparse_input</span>, </span>
<span>  ctl <span class="op">=</span> <span class="va">segList_ensemblGeneID</span><span class="op">$</span><span class="va">mouse</span><span class="op">$</span><span class="va">mouse_scSEG</span>,</span>
<span>  kmeansK <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"scMerge_hdf5"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<p>Please check out our paper for detailed analysis and results on
multiple scRNA-Seq data. <a href="https://doi.org/10.1073/pnas.1820006116" class="external-link uri">https://doi.org/10.1073/pnas.1820006116</a>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"scMerge"</span><span class="op">)</span></span>
<span><span class="co">#&gt; To cite scMerge in publications please use:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   Lin Y, Ghazanfar S, Wang K, Gagnon-Bartsch J, Lo K, Su X, Han Z,</span></span>
<span><span class="co">#&gt;   Ormerod J, Speed T, Yang P, Yang J (2019). "scMerge leverages factor</span></span>
<span><span class="co">#&gt;   analysis, stable expression, and pseudoreplication to merge multiple</span></span>
<span><span class="co">#&gt;   single-cell RNA-seq datasets." _Proceedings of the National Academy</span></span>
<span><span class="co">#&gt;   of Sciences_. doi:10.1073/pnas.1820006116</span></span>
<span><span class="co">#&gt;   &lt;https://doi.org/10.1073/pnas.1820006116&gt;,</span></span>
<span><span class="co">#&gt;   &lt;http://www.pnas.org/lookup/doi/10.1073/pnas.1820006116&gt;.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   @Article{,</span></span>
<span><span class="co">#&gt;     title = {{scMerge leverages factor analysis, stable expression, and pseudoreplication to merge multiple single-cell RNA-seq datasets}},</span></span>
<span><span class="co">#&gt;     author = {Yingxin Lin and Shila Ghazanfar and Kevin Wang and Johann Gagnon-Bartsch and Kitty Lo and Xianbin Su and Ze-Guang Han and John Ormerod and Terence Speed and Pengyi Yang and Jean Yang},</span></span>
<span><span class="co">#&gt;     year = {2019},</span></span>
<span><span class="co">#&gt;     journal = {Proceedings of the National Academy of Sciences},</span></span>
<span><span class="co">#&gt;     doi = {https://doi.org/10.1073/pnas.1820006116},</span></span>
<span><span class="co">#&gt;     url = {http://www.pnas.org/lookup/doi/10.1073/pnas.1820006116},</span></span>
<span><span class="co">#&gt;   }</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.7.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] BiocSingular_1.16.0         scater_1.28.0              </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.0               scuttle_1.10.3             </span></span>
<span><span class="co">#&gt;  [5] scMerge_1.19.0              SingleCellExperiment_1.22.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.52.1        GenomeInfoDb_1.36.4        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.34.1              S4Vectors_0.38.2           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.46.0         MatrixGenerics_1.12.3      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.3.0           BiocStyle_2.28.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3        rstudioapi_0.16.0        </span></span>
<span><span class="co">#&gt;   [3] jsonlite_1.8.8            magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [5] ggbeeswarm_0.7.2          farver_2.1.1             </span></span>
<span><span class="co">#&gt;   [7] rmarkdown_2.26            fs_1.6.3                 </span></span>
<span><span class="co">#&gt;   [9] zlibbioc_1.46.0           ragg_1.3.0               </span></span>
<span><span class="co">#&gt;  [11] vctrs_0.6.5               memoise_2.0.1            </span></span>
<span><span class="co">#&gt;  [13] DelayedMatrixStats_1.22.6 RCurl_1.98-1.14          </span></span>
<span><span class="co">#&gt;  [15] base64enc_0.1-3           htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [17] S4Arrays_1.0.6            BiocNeighbors_1.18.0     </span></span>
<span><span class="co">#&gt;  [19] Formula_1.2-5             sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [21] StanHeaders_2.32.6        reldist_1.7-2            </span></span>
<span><span class="co">#&gt;  [23] KernSmooth_2.23-22        bslib_0.7.0              </span></span>
<span><span class="co">#&gt;  [25] htmlwidgets_1.6.4         desc_1.4.3               </span></span>
<span><span class="co">#&gt;  [27] cachem_1.0.8              ResidualMatrix_1.10.0    </span></span>
<span><span class="co">#&gt;  [29] sfsmisc_1.1-17            igraph_2.0.3             </span></span>
<span><span class="co">#&gt;  [31] lifecycle_1.0.4           startupmsg_0.9.6.1       </span></span>
<span><span class="co">#&gt;  [33] pkgconfig_2.0.3           M3Drop_1.26.0            </span></span>
<span><span class="co">#&gt;  [35] rsvd_1.0.5                Matrix_1.6-5             </span></span>
<span><span class="co">#&gt;  [37] R6_2.5.1                  fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [39] GenomeInfoDbData_1.2.10   digest_0.6.35            </span></span>
<span><span class="co">#&gt;  [41] numDeriv_2016.8-1.1       colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [43] dqrng_0.3.2               irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [45] textshaping_0.3.7         Hmisc_5.1-2              </span></span>
<span><span class="co">#&gt;  [47] beachmat_2.16.0           labeling_0.4.3           </span></span>
<span><span class="co">#&gt;  [49] fansi_1.0.6               abind_1.4-5              </span></span>
<span><span class="co">#&gt;  [51] mgcv_1.9-1                compiler_4.3.3           </span></span>
<span><span class="co">#&gt;  [53] withr_3.0.0               htmlTable_2.4.2          </span></span>
<span><span class="co">#&gt;  [55] backports_1.4.1           inline_0.3.19            </span></span>
<span><span class="co">#&gt;  [57] BiocParallel_1.34.2       viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [59] highr_0.10                QuickJSR_1.1.3           </span></span>
<span><span class="co">#&gt;  [61] pkgbuild_1.4.4            gplots_3.1.3.1           </span></span>
<span><span class="co">#&gt;  [63] MASS_7.3-60.0.1           proxyC_0.4.1             </span></span>
<span><span class="co">#&gt;  [65] DelayedArray_0.26.7       bluster_1.10.0           </span></span>
<span><span class="co">#&gt;  [67] gtools_3.9.5              caTools_1.18.2           </span></span>
<span><span class="co">#&gt;  [69] loo_2.7.0                 distr_2.9.3              </span></span>
<span><span class="co">#&gt;  [71] tools_4.3.3               vipor_0.4.7              </span></span>
<span><span class="co">#&gt;  [73] foreign_0.8-86            beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt;  [75] nnet_7.3-19               glue_1.7.0               </span></span>
<span><span class="co">#&gt;  [77] batchelor_1.16.0          nlme_3.1-164             </span></span>
<span><span class="co">#&gt;  [79] cvTools_0.3.3             grid_4.3.3               </span></span>
<span><span class="co">#&gt;  [81] checkmate_2.3.1           cluster_2.1.6            </span></span>
<span><span class="co">#&gt;  [83] gtable_0.3.4              data.table_1.15.4        </span></span>
<span><span class="co">#&gt;  [85] metapod_1.8.0             ScaledMatrix_1.8.1       </span></span>
<span><span class="co">#&gt;  [87] utf8_1.2.4                XVector_0.40.0           </span></span>
<span><span class="co">#&gt;  [89] ggrepel_0.9.5             pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [91] stringr_1.5.1             limma_3.56.2             </span></span>
<span><span class="co">#&gt;  [93] robustbase_0.99-2         splines_4.3.3            </span></span>
<span><span class="co">#&gt;  [95] lattice_0.22-5            densEstBayes_1.0-2.2     </span></span>
<span><span class="co">#&gt;  [97] ruv_0.9.7.1               locfit_1.5-9.9           </span></span>
<span><span class="co">#&gt;  [99] knitr_1.46                gridExtra_2.3            </span></span>
<span><span class="co">#&gt; [101] bookdown_0.38             edgeR_3.42.4             </span></span>
<span><span class="co">#&gt; [103] xfun_0.43                 statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [105] DEoptimR_1.1-3            rstan_2.32.6             </span></span>
<span><span class="co">#&gt; [107] stringi_1.8.3             yaml_2.3.8               </span></span>
<span><span class="co">#&gt; [109] evaluate_0.23             codetools_0.2-19         </span></span>
<span><span class="co">#&gt; [111] bbmle_1.0.25.1            tibble_3.2.1             </span></span>
<span><span class="co">#&gt; [113] BiocManager_1.30.22       cli_3.6.2                </span></span>
<span><span class="co">#&gt; [115] RcppParallel_5.1.7        rpart_4.1.23             </span></span>
<span><span class="co">#&gt; [117] systemfonts_1.0.6         munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [119] jquerylib_0.1.4           Rcpp_1.0.12              </span></span>
<span><span class="co">#&gt; [121] bdsmatrix_1.3-7           parallel_4.3.3           </span></span>
<span><span class="co">#&gt; [123] rstantools_2.4.0          pkgdown_2.0.8.9000       </span></span>
<span><span class="co">#&gt; [125] scran_1.28.2              sparseMatrixStats_1.12.2 </span></span>
<span><span class="co">#&gt; [127] bitops_1.0-7              viridisLite_0.4.2        </span></span>
<span><span class="co">#&gt; [129] mvtnorm_1.2-4             scales_1.3.0             </span></span>
<span><span class="co">#&gt; [131] purrr_1.0.2               crayon_1.5.2             </span></span>
<span><span class="co">#&gt; [133] rlang_1.1.3</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yingxin Lin, Kevin Wang, Sydney Bioinformatics and Biometrics Group.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.8.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
